src/error-code.ts

```diff
@@ -760,6 +760,295 @@ export namespace ErrorCodes {
+   *
+   * Workaround - Return with the error message encoded in payload and let the machine succeed.
+   *
+   * ```ts
+   * new StepFunction<{ value: undefined }, { error?: string }>(this, 'sfn', async (input) => {
+   *    // invalid - the error cause is not constant
+   *    return {
+   *       error: input.value
+   *    }
+   * });
+   * ```
+   *
+   * TODO: add validation
+   */
+  export const StepFunctions_error_name_and_cause_must_be_constant: ErrorCode =
+    {
+      code: 10024,
+      type: ErrorType.ERROR,
+      title: "StepFunctions error name and cause must be constant",
+    };
+
+  /**
+   * Error is thrown when an array or object is incorrectly accessed.
+   *
+   * Step Functions:
+   *
+   *   1. Arrays and Objects can be accessed
+   *   2. Access Elements must be constant values - Step Functions does not support access using dynamic variables
+   *
+   * ```ts
+   * new StepFunction(stack, 'sfn', () => {
+   *    const obj = { a: "val" };
+   *    // valid
+   *    obj.a;
+   *    // valid
+   *    obj["a"];
+   *    // invalid
+   *    const a = "a";
+   *    obj[a];
+   *
+   *    const arr = [1];
+   *    // valid
+   *    arr[0]
+   *    // invalid
+   *    const zero = 0;
+   *    arr[zero];
+   * });
+   *
+   * Workaround - use lambda
+   *
+   * const accessor = new Function<{ obj: [key: string]: string, acc: string }, string>(stack, 'func', async (input) => {
+   *    return input.obj[input.acc];
+   * });
+   *
+   * new StepFunction(stack, 'sfn', () => {
+   *    // valid
+   *    await accessor(obj, a);
+   * });
+   * ```
+   */
+  export const Invalid_collection_access: ErrorCode = {
+    code: 10025,
+    type: ErrorType.ERROR,
+    title: "Invalid collect access",
+  };
+
+  /**
+   * StepFunctions does not support dynamic property names
+   *
+   * new StepFunction(stack, 'sfn', async () => {
+   *    const c = "c";
+   *    const {
+   *       a: "valid",
+   *       ["b"]: "valid",
+   *       [c]: "invalid",
+   *       [someMethod()] :"invalid"
+   *    }
+   * });
+   *
+   * Workaround - use lambda
+   *
+   * const assign = new Function<{
+   *    obj: { [key: string]: string },
+   *    key: string, value: string
+   * }>(stack, 'func', async (input) => {
+   *     return {
+   *        ...input.obj,
+   *        [input.key]: input.value
+   *     }
+   * });
+   * new StepFunction(stack, 'sfn', async () => {
+   *    return await assign({}, someMethod(), "value");
+   * })
+   */
+  export const StepFunctions_property_names_must_be_constant: ErrorCode = {
+    code: 10026,
+    type: ErrorType.ERROR,
+    title: "StepFunction property names must be constant",
+  };
+
+  /**
+   * Invalid Input
+   *
+   * Generic error code for when the user provided an unexpected input as documented and reflected in the types.
+   *
+   * See the error message for more details.
+   */
+  export const Invalid_Input: ErrorCode = {
+    code: 10027,
+    type: ErrorType.ERROR,
+    title: "Invalid Input",
+  };
```

Comment:
So, there are three ways an error can happen

1. Unexpected - 500 type error, we didn't think the error could happen
2. Unsupported - user would think it should work because it is valid ts, but our transpiler or the service doesn't support it. This is a majority of our granular errorrs.
3. Invalid Input - 400 type errors, we told the user what to send via the types, but somehow they send something else.

What should we throw in this case?

```ts
      return context.evalExpr(seconds, call, (secondsOutput) => {
        if (
          ASLGraph.isValue(secondsOutput) &&
          typeof secondsOutput.value === "number"
        ) {
          return context.voidState({
            Type: "Wait" as const,
            Seconds: secondsOutput.value,
          });
        } else if (ASLGraph.isJsonPath(secondsOutput)) {
          return context.voidState({
            Type: "Wait" as const,
            SecondsPath: secondsOutput.jsonPath,
          });
        }

        throw new SynthError(
          ???,
          "Expected the first parameter (seconds) to $SFN.waitFor to a number or a variable."
        );
      });
```

We could:
1. merge 1 and 3, all errors whether caused by the user or the library throw `Unexpected`
2. Create granular errors for any 400 type case, even when typescript should catch it. (unless ts-ignore or similar could be used.)
    3.  ex 1: "$SFN.WaitFor expected a variable or a string"
    4. ex 2: "$SFN.WaitUntil expected a variable or a number"
3. Less granular error for any 400 type cases
    3.  ex 1: "Invalid input to a step function integration"
    4. ex 2: "Invalid input to an app sync validation"
6. Create a new, generic error code for 400 type errors.
   7. Invalid Input - Message: "$SFN.WaitFor expected a variable or a string"
   8. Invalid Input - Message: "$SFN.WaitUntil expected a variable or a number"
10. Just throw ugly errors, this is ts's job, we don't think it will ever happen.

I like option 6 personally, granular will be a PITA and I like to separate 500 from 400 type errors as the messaging is much different.